### DISCUSSION TITLE
Manipulating color, filters and shaders

### DISCUSSION THREAD

**Discussion by a-1219** [brunoherbelin/vimix#152] [Q&A] [answered]
Hi, I've been exploring vimix recently and finding it to be more powerful than its simple interface shows. I have a few questions (that may also be suggestions):

Is there a way to work with color correction and filters on the main output?

Is it possible to permanently import custom shaders to vimix (as if they we're built-in, not just for one session) and edit the code of native ones? Where can I find their files on Linux?

Thanks in advance!

**Comment by brunoherbelin** [comment]
Hi! sorry it took long to reply! Busy times...

> Is there a way to work with color correction and filters on the main output?

Yes, this is done in the `Displays` view `F5` ; 

![screenshot](https://user-images.githubusercontent.com/60393189/228677655-13bf9ce0-547e-4855-aef5-138aaa1a5162.png)

> Is it possible to permanently import custom shaders to vimix 

You mean shaders that would be listed in a user defined list, so that they appear in the selection? 
Hum.. could be a good idea! I will try to implement this ! 

> ... and edit the code of native ones? Where can I find their files on Linux?

Yes, the GLSL code is in https://github.com/brunoherbelin/vimix/tree/master/rsc/shaders/filters

**Comment by a-1219** [comment]
Don't worry!

> Yes, this is done in the Displays view F5 ;

I was thinking about being able to also add filters to the whole session and to control the output brightness and contrast levels, somewhat like in the Advanced Output in Resolume for instance. I'm currently blending a white shader using the screen mode to adjust the overall brightness.

> You mean shaders that would be listed in a user defined list, so that they appear in the selection?

Exactly!

>... and edit the code of native ones? Where can I find their files on Linux?

It would be nice to edit their code directly on Vimix. I tried to find the .glsl files on my computer (I'm using the flatpak version) without success. Have you considered implementing ISF shaders?

Sorry if these questions are out of the scope of the project, I know it's a lot of work.

**Comment by brunoherbelin** [comment]
Hey there! Just implemented sliders for contrast and brightness of output windows :

![20241123100746327_vmixcapture](https://github.com/user-attachments/assets/a1ad31d4-eb99-48d1-a2d9-c3ddc95e1232)

Available with Beta commit ce4e964f0940076ed538e40de8f2039453370575

**Reply by a-1219** [reply]
That's awesome!

**Comment by brunoherbelin** [comment]
Work in progress: added to Shader Editor the possibility to refer to text files with GLSL code (aka shader toy style) for custom shader filters. 
TODO : automatically rebuild all sources that uses the same shader file when changing the code.

Thanks for trying Beta commit  df0913727c7ad2da275fb5b67dbb9dce8f6a2940 and letting me know !

**Reply by lupin3rd** [reply]
Thanks @brunoherbelin ! Very usefull!
It might be useful to be able to insert a source created by the shader editor in insert->patterns & generated graphics.
We don't always use GLSL code to effect on existing media. Sometimes we use it to generate graphics.
What do you think?

**Reply by brunoherbelin** [reply]
Yep, I finally found a moment to finish this
Now there is a new type of sources 'Shader' which generates graphics from GLSL code.

![20250223230318300_vmixcapture](https://github.com/user-attachments/assets/8340e167-d544-4d99-ad46-65f91ecf7ed1)


To be tested in Beta d01366d0690804711b683f951156254247ca17e2

**Reply by lupin3rd** [reply]
I try the last beta... i can create a custom shader source... but if i open editor i can't edit or use embedded code.

**Comment by brunoherbelin** [comment, marked as answer]
![Screenshot From 2024-12-15 18-37-34](https://github.com/user-attachments/assets/44883a01-db57-454c-aa46-f888e9c418c8)

Okay, so with commit 0289f4c06ed8c343be8b30f768922ff1dbe860b9, things seems to be operational

The principle is simple: if the shader is selected from a file (txt file *.glsl) then the file is used as code for the shader (saved when build, loaded from file, etc.), otherwise the code is embedded as before.

If you change the code from outside, and multiple filters use the same shader file, use the menu entry 'Build all'.
